---
title: ALVACComp gene-expression analysis
author: Slim Fourati
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: github_documents
---

Loading require packages
```{r, loading-packages}
suppressPackageStartupMessages(library(package = "knitr"))
suppressPackageStartupMessages(library(package = "parallel"))
suppressPackageStartupMessages(library(package = "Biobase"))
suppressPackageStartupMessages(library(package = "limma"))
suppressPackageStartupMessages(library(package = "impute"))
suppressPackageStartupMessages(library(package = "org.Hs.eg.db"))
suppressPackageStartupMessages(library(package = "pheatmap"))
suppressPackageStartupMessages(library(package = "grid"))
suppressPackageStartupMessages(library(package = "gtable"))
suppressPackageStartupMessages(library(package = "tidyverse"))
```

Set default options/variables
```{r, global-variables}
workDir <- dirname(getwd())
opts_chunk$set(tidy = FALSE, fig.path = "../figure/")
options(stringsAsFactors  = FALSE,
        width             = 80,
        mc.cores          = detectCores() - 1,
        readr.num_columns = 0)
```

Read non-normalized matrix
```{r, read-raw-matrix, warning=FALSE}
rawFile <- file.path(workDir,
		     "input",
		     "GA_illumina_expression.alvaccomp.matrix_non_norm.csv")
rawMat <- read_csv(file = rawFile, progress = FALSE)
```

Read arrays annotation
```{r, reading-arrays-annotation}
arraysAnnotFile <- file.path(workDir,
			     "input",
			     "GA_illumina_expression.alvaccomp.metadata.csv")
arraysAnnotation <- read_csv(file = arraysAnnotFile, progress = FALSE)
# remove unused phenotypic information
arraysAnnotation <- select(arraysAnnotation,
                           -title,
                           -`source name`,
                           -organism,
                           -molecule,
                           -label,
                           -description,
                           -platform)
# remove prefix 'characteristics` of column names
setNames(arraysAnnotation, nm = gsub(pattern = "^[^:]+: (.+)$",
                                     replacement = "\\1",
                                     names(arraysAnnotation))) ->
  arraysAnnotation
```

Read features annotation
```{r, reading-features-annotation}
# featuresAnnotFile <- file.path(workDir,
# 			       "input/Illumina_HumanHT12_V4.rheMac3.chip")
# featuresAnnotation <- read_tsv(file = featuresAnnotFile, progress = FALSE) %>%
#     as.data.frame()
# rownames(featuresAnnotation) <- featuresAnnotation$IlmnID

# featuresAnnotFile <- file.path(workDir,
#                                "input/Illumina_HumanHT12_V4.rheMac3.chip")
#                                
                               
featuresAnnotFile <- file.path(workDir, "../input/featureAnnotation_study36.csv")

# featuresAnnotation <- read_tsv(file = featuresAnnotFile, progress = FALSE) %>%
#   as.data.frame()
  
featuresAnnotation <- read_csv(file = featuresAnnotFile, progress = FALSE) %>%
  as.data.frame() %>%
  column_to_rownames("X1")
# rownames(featuresAnnotation) <- featuresAnnotation$IlmnID
```

Create non-normalized ExpressionSet
```{r, create-raw-eset}

# format raw matrix
rawMat <- read_csv(file = rawFile, progress = FALSE)
rNames <- rawMat$"ID_REF"

rawMat <- rawMat[, -grep(pattern = "ID_REF|Detection Pval",
                         colnames(rawMat))]


rawMat <- as.matrix(rawMat)
rownames(rawMat) <- rNames
# format phenodata
arraysAnnotation <- as.data.frame(arraysAnnotation)
rownames(arraysAnnotation) <- arraysAnnotation$"Sample name"

arraysAnnotation <- arraysAnnotation[colnames(rawMat), ]
# format feature annotation
featuresAnnotation <- as.data.frame(featuresAnnotation)


assayRownames <- intersect(row.names(rawMat), row.names(featuresAnnotation))



rawMat <- rawMat[row.names(rawMat) %in% assayRownames,]
featuresAnnotation <- featuresAnnotation[row.names(featuresAnnotation) %in% assayRownames, ]
featuresAnnotation <- featuresAnnotation[rownames(rawMat), ]
#
# # create ExpressionSet
esetRaw <- ExpressionSet(assayData   = rawMat,
                         phenoData   = AnnotatedDataFrame(arraysAnnotation),
                         featureData = AnnotatedDataFrame(featuresAnnotation))
# save raw ExpressionSet
save(esetRaw, file = file.path(workDir, "output/alvaccomp.esetRaw.RData"))
```

Quality control: kernel densities plot
```{r, density-plot, fig.width=8, fig.height=4}
intensities <- as.vector(exprs(esetRaw))
# log2 transform (replace intensities < 1 by 1 to prevent -Inf)
intensities <- log2(pmax(intensities, 1))
arrays <- rep(sampleNames(esetRaw), each = nrow(esetRaw))
plotDF <- data.frame(intensities, arrays)
ggplot(data    = plotDF,
       mapping = aes(x     = intensities,
		     color = factor(arrays))) +
  scale_color_discrete(name = "scanning note") +
  stat_density(geom = "path", position = "identity") +
  labs(x     = "log2 intensity",
       title = "density plot [log2 raw intensities]") +
  theme_bw() +
  theme(legend.position = "none")
```

Quality control: multidimentional scaling plot
```{r, mds-plot, fig.width=6.5, fig.height=5}
# create tempory ExpressionSet and normalize the data
esetTemp <- esetRaw
rawMat <- exprs(esetTemp)
normMat <- normalizeBetweenArrays(rawMat, method = "quantile")
normMat <- log2(normMat)
# generate a multidimensional scaling plot
distMat <- dist(t(normMat))
mds <- cmdscale(d = distMat, k = ncol(normMat) - 1, eig = TRUE)
mdsStDev <- apply(mds$points, MARGIN = 2, FUN = sd)
mdsPercVar <- round((mdsStDev^2)/sum(mdsStDev^2) * 100)
tecRep <- grep(pattern = "_rep1", esetRaw$"Sample name", value = TRUE) %>%
  gsub(pattern = "_rep1", replacement = "")
dupArrays <- gsub(pattern = "_rep1", replacement = "", esetRaw$"Sample name")
dupArrays[!dupArrays %in% tecRep] <- NA
plotDF <- data.frame(V1               = mds$points[, 1],
                     V2               = mds$points[, 2],
                     `technical rep.` = dupArrays,
                     check.names      = FALSE)
ggplot(data = plotDF,
       mapping = aes(x = V1, y = V2, color = `technical rep.`)) +
  geom_point(size = 3) +
    labs(x     = paste0("1st dimension (", mdsPercVar[1], "%)"),
	 y     = paste0("2nd dimension (", mdsPercVar[2], "%)"),
	 title = "Multidimentional scaling plot") +
    theme_bw() +
    theme(legend.key = element_blank())
```

Normalizing raw expression
```{r, normalization}
eset <- esetRaw
# order esetRaw by idat file name and features by ProbeID
eset <- eset[order(as.numeric(fData(eset)$ProbeID)),
             order(eset$"idat file")]
# impute missing intensities (intensities = 0)
rawMat <- exprs(eset)
rawMat[rawMat == 0] <- NA
suppressWarnings(capture.output(rawMat <- impute.knn(data = rawMat)$data,
                                file = "/dev/null"))
exprs(eset) <- rawMat
# quantile normalized and log2 transform expression
normMat <- normalizeBetweenArrays(exprs(eset), method = "quantile")
# surrogate remplacement
normMat[normMat < 2^0.1] <- 2^0.1
normMat <- log2(normMat)
exprs(eset) <- normMat
# merge techinical replicates
exprsMat <- exprs(eset) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  gather(`Sample name`, value, -rowname) %>%
  mutate(`Sample name` = gsub(pattern     = "_rep1",
                              replacement = "", 
                              `Sample name`)) %>%
  group_by(`Sample name`, rowname) %>%
  summarise(mu = mean(value)) %>%
  spread(`Sample name`, mu) %>%
  as.data.frame()
rownames(exprsMat) <- exprsMat$rowname
exprsMat$rowname <- NULL
eset <- eset[, colnames(exprsMat)]
exprs(eset) <- as.matrix(exprsMat)[featureNames(eset), ] 
# save normalized ExpressionSet
save(eset, file = file.path(workDir, "output/alvaccomp.eset.RData"))
```

quality control: gender check
```{r, boxplot-chry, fig.width=3, fig.height=4}
symbol2chr <- merge(as.data.frame(org.Hs.egSYMBOL),
                    as.data.frame(org.Hs.egCHR),
                    by = "gene_id")
pb2symbol <- strsplit(fData(eset)$SYMBOL, split = " /// ") %>%
  setNames(fData(eset)$IlmnID) %>%
  stack() %>%
  mutate(ind = as.vector(ind))
pb2chr <- merge(symbol2chr, pb2symbol, by.x = "symbol", by.y = "values")

pbY <- filter(pb2chr, chromosome %in% "Y") %>%
  .$ind

plotDF <- data.frame(mu = colMeans(exprs(eset)[pbY, ])) %>%
  rownames_to_column() %>%
  merge(pData(eset), by.x = "rowname", by.y = "Sample name")

ggplot(data = plotDF, mapping = aes(x = donor, y = mu)) +
  geom_jitter() +
  labs(y = "Average expression of chr.Y probes") +
  theme_bw()
```

Exploratory analysis: heatmap based on top 100 most varying transcripts
```{r, exploratory-heatmap}
bluered <- colorRampPalette(colors = c("blue", "white", "red"))
varList <- apply(exprs(eset), MARGIN = 1, FUN = var)
esetTemp <- eset[order(varList, decreasing = TRUE)[1:100], ]
esetTemp$donor <- factor(esetTemp$donor)
exploHeat <- pheatmap(mat            = exprs(esetTemp),
                      color          = bluered(100),
                      scale          = "row",
                      treeheight_row = 0,
                      annotation_col = pData(esetTemp)[,
                          c("donor",
                            "vaccination time",
                            "time after immunization")],
                      show_colnames  = FALSE,
                      show_rownames  = FALSE,
                      main           = paste0("top 100 varying probes",
                          "[Euclidean distance, complete linkage]"),
                      silent         = TRUE)
colorName <- textGrob("z-score", x = 0.5, y = 1.05, gp = gpar(fontface = "bold"))
exploHeat$gtable <- gtable_add_grob(exploHeat$gtable,
                                    colorName,
                                    t    = 3,
                                    l    = 5,
                                    b    = 5,
                                    clip = "off",
                                    name = "colorName")
grid.draw(exploHeat$gtable)
# test association with time after immunization
tab <- table(cutree(exploHeat$tree_col, k = 2),
             esetTemp$"time after immunization")
fit <- fisher.test(tab)
print(tab)
print(fit)
# test for clustering of samples by participants
participantList <- esetTemp$donor[exploHeat$tree_col$order]
y <- sum(participantList[-length(participantList)] == participantList[-1])
# derive p-value using a 1000 fold permutation test
B <- 1000
yhat <- mclapply(1:B, function(seed) {
  set.seed(seed = seed)
  participantList <- sample(esetTemp$donor)
  return(value = sum(participantList[-length(participantList)] ==
             participantList[-1]))
})
print(paste0("permutation test: p<=", max(mean(unlist(yhat) >= y), 1/B)))
```

Create a prevax-substracted ExpressionSet
```{r, substracted-eset}
# identify complete pair of ENV-DMSO stimulated samples
esetPre <- eset[, eset$"vaccination time" %in% "pre-vaccination"]
esetBaselined <- eset[, eset$"vaccination time" != "pre-vaccination"]
exprs(esetBaselined) <- exprs(esetBaselined) - 
  exprs(esetPre[, match(esetBaselined$donor, table = esetPre$donor)])
# save baselined expression
save(esetBaselined, file = file.path(workDir, "output/alvaccomp.esetBaselined.RData"))
```

```{r, deg-stim-casectrl, echo=FALSE, eval=FALSE}
#Differential expression analysis
# create list to save linear models
fits <- list()
# identify samples stimulated both by ENV and DMSO
flag <- pData(eset) %>%
        select(`Sample name`, donor, stimulation) %>%
        spread(stimulation, `Sample name`) %>%
        filter(!is.na(`92TH023ENV`) & !is.na(DMSO)) %>%
        gather(stimulation, `Sample name`, -donor)
esetTemp <- eset[, flag$"Sample name"]
# identify genes differently expressed separetly for the VACCINE and PLACEBO
# treatment groups
TreatmentStim <- interaction(esetTemp$treatment,
                             esetTemp$stimulation,
                             sep = "_")
design <- model.matrix(~0+TreatmentStim)
colnames(design) <- gsub(pattern     = "TreatmentStim",
                         replacement = "",
                         colnames(design))
rownames(design) <- sampleNames(esetTemp)
# fit a fixed effect to the treatment/strim and a random effect to the donor in
# the linear regression model
donor <- factor(esetTemp$donor)
corfit <- duplicateCorrelation(exprs(esetTemp),
                               design = design,
                               block  = esetTemp$donor)
fit <- lmFit(esetTemp,
             design      = design,
             block       = donor,
             correlation = corfit$consensus)
contrastLS <- c("PLACEBO_92TH023ENV-PLACEBO_DMSO", "VACCINE_92TH023ENV-VACCINE_DMSO")
contrastMat <- makeContrasts(contrasts = contrastLS, levels = fit$design)
fit2 <- contrasts.fit(fit = fit, contrasts = contrastMat)
fit2 <- eBayes(fit = fit2)
fits[["stimulation"]] <- list(fit = fit, fit2 = fit2)

# differential expression analysis: HIV case vs. control
TreatmentInfect <- interaction(esetBaselined$treatment,
                               esetBaselined$"HIV infection",
                               sep       = "_",
                               lex.order = TRUE)
design <- model.matrix(~0+TreatmentInfect)
colnames(design) <- gsub(pattern     = "TreatmentInfect",
                         replacement = "",
                         colnames(design))
rownames(design) <- sampleNames(esetBaselined)
fit <- lmFit(esetBaselined, design = design)
contrastLS <- c("PLACEBO_No-PLACEBO_Yes",
                "VACCINE_No-VACCINE_Yes")
contrastMat <- makeContrasts(contrasts = contrastLS, levels = fit$design)
fit2 <- contrasts.fit(fit = fit, contrasts = contrastMat)
fit2 <- eBayes(fit = fit2)
# top 10 differently expressed transcript in the placebo group
topTable(fit = fit2, coef = 1) %>%
  select(SYMBOL, logFC, t, P.Value, adj.P.Val) %>%
  print()
# top 10 differently expressed transcript in the vaccine group
topTable(fit = fit2, coef = 2) %>%
  select(SYMBOL, logFC, t, P.Value, adj.P.Val) %>%
  print()
# save MArrayLM in list
fits[["infection"]] <- list(fit = fit, fit2 = fit2)

#save MArrayLM list
save(fits, file = file.path(workDir, "output/alvaccomp.fits.RData"))
```

```{r, heatmap-top50-stim, fig.width=8, fig.height=6, echo=FALSE, eval=FALSE}
modelName <- "stimulation"
fit2 <- fits[[modelName]][["fit2"]]
for (coefName in colnames(fit2)) {
  coefLS <- unlist(strsplit(coefName, split = "-")) %>%
      gsub(pattern = "\\(|\\)", replacement = "")
  sampleLS <- fit2$design %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    select_(.dots = c("rowname", coefLS)) %>%
    filter(apply(., MARGIN = 1, FUN = function(x) any(x %in% 1)))
    top <- topTable(fit    = fit2,
                  coef   = coefName,
                  number = Inf) %>%
         mutate(SYMBOL = gsub(pattern     = "^([^ ]+) ///.+",
                              replacement = "\\1",
                              SYMBOL)) %>%
         filter(SYMBOL != "---") %>%
         top_n(n = 50, wt = abs(t))
  exprsMat <- exprs(eset)[top$IlmnID, sampleLS$rowname] %>%
    (function(x) return(value = t(scale(t(x)))))
  breakLS <- c(-1 * max(abs(exprsMat)),
               seq(from       = -1 * min(abs(range(exprsMat))),
                   to         = min(abs(range(exprsMat))),
                   length.out = 99),
               max(abs(exprsMat)))
  colAnnotDF <- pData(eset) %>%
    .[, c("stimulation", "treatment")]
  rowAnnotDF <- top %>%
    mutate(`fold-change sign` = ifelse(test = sign(logFC) %in% 1,
                                       yes  = "UP",
                                       no  = "DN"),
           `-log10 P.Value` = -1 * log10(P.Value),
           `adj. p <= 0.05` = factor(adj.P.Val <= 0.05)) %>%
    select(IlmnID, `fold-change sign`, `-log10 P.Value`, `adj. p <= 0.05`)
  rownames(rowAnnotDF) <- rowAnnotDF$IlmnID
  rowAnnotDF$IlmnID <- NULL
  colorAnnotLS <- list(treatment = c(PLACEBO = "blue", VACCINE = "green"),
                       "fold-change sign" = c(UP = "red", DN = "blue"))
  pheat <- pheatmap(mat               = exprsMat,
                    color             = bluered(100),
                    breaks            = breakLS,
                    cellwidth         = 2,
                    cellheight        = 2,
                    treeheight_row    = 0,
                    annotation_col    = colAnnotDF,
                    annotation_row    = rowAnnotDF,
                    annotation_colors = colorAnnotLS,
                    show_colnames     = FALSE,
                    main              = coefName,
                    fontsize_row      = 2,
                    labels_row        = top$SYMBOL)
}
``` 

```{r, heatmap-top50-infection, fig.width=8, fig.height=6, echo=FALSE, eval=FALSE}
modelName <- "infection"
fit2 <- fits[[modelName]][["fit2"]]
for (coefName in colnames(fit2)) {
  coefLS <- unlist(strsplit(coefName, split = "-")) %>%
      gsub(pattern = "\\(|\\)", replacement = "")
  sampleLS <- fit2$design %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    select_(.dots = c("rowname", coefLS)) %>%
    filter(apply(., MARGIN = 1, FUN = function(x) any(x %in% 1)))
    top <- topTable(fit    = fit2,
                  coef   = coefName,
                  number = Inf) %>%
         mutate(SYMBOL = gsub(pattern     = "^([^ ]+) ///.+",
                              replacement = "\\1",
                              SYMBOL)) %>%
         filter(SYMBOL != "---") %>%
         top_n(n = 50, wt = abs(t))
  exprsMat <- exprs(eset)[top$IlmnID, sampleLS$rowname] %>%
    (function(x) return(value = t(scale(t(x)))))
  breakLS <- c(-1 * max(abs(exprsMat)),
               seq(from       = -1 * min(abs(range(exprsMat))),
                   to         = min(abs(range(exprsMat))),
                   length.out = 99),
               max(abs(exprsMat)))
  colAnnotDF <- pData(eset) %>%
    .[, c("treatment", "HIV infection")]
  rowAnnotDF <- top %>%
    mutate(`fold-change sign` = ifelse(test = sign(logFC) %in% 1,
                                       yes  = "UP",
                                       no  = "DN"),
           `-log10 P.Value` = -1 * log10(P.Value),
           `adj. p <= 0.05` = factor(adj.P.Val <= 0.05)) %>%
    select(IlmnID, `fold-change sign`, `-log10 P.Value`, `adj. p <= 0.05`)
  rownames(rowAnnotDF) <- rowAnnotDF$IlmnID
  rowAnnotDF$IlmnID <- NULL
  colorAnnotLS <- list(treatment = c(PLACEBO = "blue", VACCINE = "green"),
                       "HIV infection" = c(No = "black", Yes = "red"),
                       "fold-change sign" = c(UP = "red", DN = "blue"))
  pheat <- pheatmap(mat               = exprsMat,
                    color             = bluered(100),
                    breaks            = breakLS,
                    cellwidth         = 4,
                    cellheight        = 4,
                    treeheight_row    = 0,
                    annotation_col    = colAnnotDF,
                    annotation_row    = rowAnnotDF,
                    annotation_colors = colorAnnotLS,
                    show_colnames     = FALSE,
                    main              = coefName,
                    fontsize_row      = 4,
                    labels_row        = top$SYMBOL)
}
``` 

print session info
```{r, session-info}
sessionInfo()
```

