---
title: "Study_chip"
author: "Adam Pelletier"
date: "12/1/2020"
output: html_document
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(rmarkdown))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(broom))
suppressPackageStartupMessages(library(Biobase))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ChIPseeker))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(biomaRt))
suppressPackageStartupMessages(library(VennDiagram))
suppressPackageStartupMessages(library(GSEABase))
suppressPackageStartupMessages(library(SuperExactTest))
suppressPackageStartupMessages(library(rlist))
library("TxDb.Mmulatta.UCSC.rheMac10.refGene")
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(getwd())
```



```{r }
out <- data %>% 
       separate(Date, into = c("Year", "Month", "Day"), sep = "-") %>%
       mutate(Day = as.numeric(Day)) %>%
       gather(Stock, Stock_value, -Day, -Month, -Year) %>%
       group_by(Month) %>%
       filter(Day == max(Day)) %>%
       mutate(decile = ntile(Stock1, n = 10)) %>%
       ungroup() %>%
       select(-Stock_value) %>%
       spread(Stock, decile)


```

```{r read_peaks}

peak_files <- list("Human" = 
       list("Empty" = "Human/Human_empty_vs_media_24hr/peak/overlap_reproducibility/overlap.conservative_peak.narrowPeak.gz",
            "Full" = "Human/Human_full_vs_media_24hr/peak/overlap_reproducibility/overlap.optimal_peak.narrowPeak.gz"))


peak_calls <- sapply(names(peak_files), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_files[[x]]), simplify = F, USE.NAMES = T, function(y){
     out <- readPeakFile(peak_files[[x]][[y]])
   })
  
})

covplots <- sapply(names(peak_calls), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_calls[[x]]), simplify = F, USE.NAMES = T, function(y){
    out <- covplot(peak_calls[[x]][[y]], weightCol = "V7")
  })
})

jpeg("output/figures/Human_full_alvac_vs_media.jpg")
covplots$Human$Full
dev.off()
```





```{r promoters}



txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene


promoter <- getPromoters(TxDb=txdb, upstream=1000, downstream=1000)

annot_matrix <- sapply(names(peak_calls), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_calls[[x]]), simplify = F, USE.NAMES = T, function(y){
    out <- annotatePeak(peak_calls[[x]][[y]], TxDb =txdb, tssRegion = c(-3000,3000),
                        annoDb="org.Hs.eg.db") 
  })
})




annot_df_promoter <- sapply(names(annot_matrix), simplify = F, USE.NAMES = T, function(x){
    peak <- do.call("rbind", lapply(names(annot_matrix[[x]]), function(y){
      out <- as.data.frame(annot_matrix[[x]][[y]]) %>%
              mutate(species = x) %>%
              mutate(condition = y) %>%
              filter(grepl("Promoter", annotation)) %>%
          filter(grepl(paste(immune_filter, collapse = "|"),
                           SYMBOL))
  })) #%>%
  #group_by(annotation, SYMBOL) %>%
  #mutate(n_rep = n_distinct(condition)) %>%
  #ungroup() #%>%
  #filter(n_rep == max(n_rep)) #%>%
  #filter(n_rep == 2)# %>%
  #filter(grepl("CCL|CXCL|CX3CL|DDX|TMEM", SYMBOL))
}) 





prom_bed_full <- as.data.frame(annot_matrix$Human$Full) %>%
              filter(grepl("Promoter", annotation))# %>% 
              #filter(distanceToTSS <0) %>%


mart = useMart("ensembl", dataset="hsapiens_gene_ensembl")


ensembl_id <-  getBM(attributes= c("hgnc_symbol","ensembl_gene_id", "strand"),
                  filters = 'hgnc_symbol', values = prom_bed_full$SYMBOL, mart = mart) 


filt_bed_full <- prom_bed_full %>%
              dplyr::select(-strand) %>%
              #left_join(., ensembl_id, by = c("SYMBOL" = "hgnc_symbol")) %>%
              filter(distanceToTSS < 300) %>%
              filter(distanceToTSS > -3000) %>%
              dplyr::select(seqnames, start, end, V7, SYMBOL, V4, V9)


qc_human_peaks <- filt_bed_full %>%
          filter( V9 > -log2(0.05)) %>%
          #filter(V7 >5 & V9 > -log2(0.05)) %>%
          summarise(n = n()) 
      


write.table(filt_bed_full %>% dplyr::select(-SYMBOL, -V4), 
            file = "Human/Human_full_vs_media_24hr/peak/overlap_reproducibility/filt_promoter_fc_pooled.bed", quote = F, 
            col.names = F, row.names = F)


filt_bed_full_qvals <- prom_bed_full%>%
              dplyr::select(-strand) %>%
              #left_join(., ensembl_id, by = c("SYMBOL" = "hgnc_symbol")) %>%
              filter(distanceToTSS < 300) %>%
              filter(distanceToTSS >-3000) %>%
              # inner_join(., ensembl_id_NHP, by = c("transcriptId" = "refseq_mrna")) %>%
              # inner_join(., hsapiens_orthologs, by  = "ensembl_gene_id") %>%
              # dplyr::rename(SYMBOL = "hsapiens_homolog_associated_gene_name") %>%
              # filter(SYMBOL != "") %>%
              dplyr::select(seqnames, start, end, V9, SYMBOL, V4) %>%
              mutate(qval = 10^(-V9)) %>%
              filter(SYMBOL %in% c("FLT3LG", "TBK1", "HLA-DMB")) 
  

write_tsv(filt_bed_full_qvals, path = "output/suppl5E_peak_qvalues_human.txt")


peak_calls_filt <- peak_calls$Human$Full[peak_calls$Human$Full$V4 %in% filt_bed_full$V4,]

covplot(peak_calls_filt, weightCol = "V7")


rv144_hiv_acq <- readLines("../transcriptomicss/rv144/output/data/CREB_hiv_acquisition_LE.txt")
rv144_induced <- readLines("../transcriptomicss/rv144/output/data/CREB1_postvaxx_signature.txt")
rv144_titers <- readLines("../transcriptomicss/rv144/output/data/CREB_rv144_titers_LE.txt")

CREB_CHEA <- getGmt("~/Documents/GMT_files/CHEA-tf.gmt")
CREB_CHEA_LE <- geneIds(CREB_CHEA[names(CREB_CHEA) == "CREB1"])[[1]]



putative_creb_raw <- read_excel("../transcriptomicss/input/01076Table3.xls", skip = 1) %>%
                      separate_rows(`Alias Symbols`, sep = ";")

alias <- putative_creb_raw$`Alias Symbols`[!is.na(putative_creb_raw$`Alias Symbols`)]
mart_m = useMart("ensembl", dataset="mmusculus_gene_ensembl")
# putat_orth_m <-  getBM(attributes= c("ensembl_gene_id", "external_gene_name"),
#                   filters = "external_gene_name", values = putative_creb_raw$Symbol, mart = mart_m)

putat_orth <-  getLDS(attributes= c("external_gene_name"),
                                   
                  filters = "external_gene_name", 
                  values = putative_creb_raw$Symbol, 
                             mart = mart_m,
                  attributesL= c("external_gene_name"),martL = mart) 



venn.diagram(list("CHIP-Seq\nCREB1 Peaks" = filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)],
                  "CHEA CREB1" = CREB_CHEA_LE),
             filename = "output/figures/venn_diagram_Full ALVAC_chip_vs_CHEA.tiff",
              fill = c(4,5),
             margin = 0.15, width = 3, height = 3, units = "in", resolution = 300,
             cex = 0.5, cat.cex = 0.5, cat.dist = c(0.1,0.15), cat.pos = c(225,135))


c7_gs <- getGmt("~/Documents/MSigDB/msigdb_v7.2_GMTs/c7.all.v7.2.symbols.gmt")


immune_filter <- c('CCL', 'CXCL', 'CX3CL', "CCR", 'CXCR', "CX3CR", "MIP",
                   'DDX', 'TMEM', 'IL[0-9]+', 'IFN', 'TGF' , 'FLT3L', 'TBK', 'TNF',
                   '^CD[0-9]+', 'IRF', 'CDKN', 'HLA', 'FOS', 'JUN', 'TLR', 'MAPK', 'NOD', 'NLR',
                   'NFKB', 'REL[ABC]', "STAT")




sign_gene_list <- list("RV144 HIV Acquisition" = rv144_hiv_acq,
                  "RV144 Induced Post-Vaccination" = rv144_induced,
                  "RV144 Wk26 IgG titers" = rv144_titers,
                  "CHIP-Seq CREB1 Peaks" = filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)])

venn.diagram(list("RV144 HIV\nAcquisition" = unique(rv144_hiv_acq),
                  "RV144 Induced\nPost-Vaccination" = unique(rv144_induced),
                  "RV144 Wk26\nIgG titers" = unique(rv144_titers),
                  "CHIP-Seq\nCREB1 Peaks" = unique(filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)])),
             filename = "output/figures/venn_diagram_Full ALVAC_chip.tiff",
              fill = c(5,2,3,4),
             margin = 0.25, width = 3, height = 3, units = "in", resolution = 300,
             cex = 0.5, cat.cex = 0.5, cat.dist = 0.20, 
             cat.just = list(c(1,0),c(0,0),c(0.5,1),c(0.5,1)))

raw_contrasts <- do.call("rbind", lapply(names(sign_gene_list)[c(1:3)], function(x){
   out <- data.frame("SYMBOL" = intersect(sign_gene_list[[x]],  
                                          sign_gene_list[["CHIP-Seq CREB1 Peaks"]]),
                     "condition" = x)
})) %>%
  group_by(SYMBOL) %>%
  mutate(n_conditions  =  n_distinct(condition)) %>%
  mutate(conditions = paste(condition, collapse = "; ")) %>%
  ungroup() %>%
  dplyr::select(-condition) %>%
  unique() %>%
  mutate(immune_filter = ifelse(grepl(paste(immune_filter, collapse = "|"),SYMBOL), 1, 0)) %>%
  arrange(desc(n_conditions), desc(immune_filter)) %>%
  dplyr::select(SYMBOL, immune_filter, everything())
  
writexl::write_xlsx(raw_contrasts, path = "output/rv144_gene_contrasts_raw.xlsx")

contrasts_rv144_annot <- read_xlsx("output/rv144_gene_contrasts_raw_JAT.xlsx") %>%
                            filter(immune_filter == 1)

test2 <- test[grepl(paste(immune_filter, collapse = "|" ), test )]
test2 <- rv144_hiv_acq[grepl(paste(immune_filter, collapse = "|" ), rv144_hiv_acq )]
test2 <- rv144_induced[grepl(paste(immune_filter, collapse = "|" ), rv144_induced )]
test2 <- rv144_titers[grepl(paste(immune_filter, collapse = "|" ), rv144_titers )]
test2 <- sign_gene_list$`CHIP-Seq CREB1 Peaks`[grepl(paste(immune_filter, collapse = "|" ), sign_gene_list$`CHIP-Seq CREB1 Peaks` )]


signature_overlap <- Reduce(intersect, 
                            list(rv144_induced, 
                                 rv144_hiv_acq, 
                                 rv144_titers, 
                               filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)],
                               unique(unlist(geneIds(c7_gs), use.names = F))))

writeLines(raw_contrasts[raw_contrasts$immune_filter == 1,]$SYMBOL, 
           con = "output/signatures_overlap_genes")

```



```{r study36_overlap}

study36_gsea_file <- "../transcriptomicss/output/data/CREB1_target_genes_study36.xlsx"
study_36_genes <- do.call('rbind', lapply(excel_sheets(study36_gsea_file), function(x){
            df <- read_xlsx(study36_gsea_file, sheet = x) %>%
                  separate_rows(conditions, sep = "; ") %>%
                  mutate(celltype = gsub("G[12]_", "", conditions)) %>%
                  mutate(celltype = gsub("_wk.*", "", celltype)) %>%
                  dplyr::select(-conditions) %>%
                  unique()
} )) %>%
  group_by(CREB_target) %>%
  mutate(n_contrasts = n_distinct(contrast)) %>%
  # group_by(CREB_target, celltype) %>%
  # mutate(n_contrasts = n_distinct(contrast)) %>%
  ungroup() %>%
  # filter(grepl(paste(immune_filter, collapse = "|"), CREB_target)) %>%
  # dplyr::select(CREB_target, contrast) %>% unique()
  # 
  
  dplyr::select(CREB_target, n_contrasts, celltype) %>%
  unique() %>%
  #filter(n_contrasts == 2) %>%
  mutate(immune_filter = ifelse(grepl(paste(immune_filter, collapse = "|"), CREB_target), 1, 0)) %>%
  dplyr::rename(SYMBOL = "CREB_target" ) %>%
  group_by(SYMBOL) %>%
  mutate(n_cell = n_distinct(celltype)) %>%
  mutate(celltypes = paste(celltype, collapse = "; ")) %>%
  ungroup() %>%
  mutate(immune_filter = ifelse(SYMBOL %in% contrasts_rv144_annot$SYMBOL, 1, immune_filter)) %>%
  arrange(desc(n_cell), desc(immune_filter)) 
  #unique() 
  #split(., .$celltype)

study_36_contrasts <- list("Study36 B" = unique(study_36_genes[study_36_genes$celltype ==
                                                           "Bcells",]$SYMBOL),
                  "Study36 T" = unique(study_36_genes[study_36_genes$celltype == "Tcells",]$SYMBOL),
                  "Study36 DC" = unique(study_36_genes[study_36_genes$celltype == "DCcells",]$SYMBOL),
                  "CHIP-SeqCREB1 Peaks" = filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)])

venn.diagram(study_36_contrasts,
             filename = "output/figures/venn_diagram_study36__Full ALVAC_chip.tiff",
              fill = c(5,2,3,4),
             margin = 0.15, width = 3, height = 3, units = "in", resolution = 300,
             cex = 0.5, cat.cex = 0.5)

writexl::write_xlsx((study_36_genes %>%
                      dplyr::select(-celltype) %>% unique(.)) %>%
                   
                      filter(SYMBOL %in% filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)]),
  #dplyr::select(-celltype) %>%,
                       path = "output/study36_gene_contrasts_raw.xlsx")



```


```{r venn_test}

extract_superset_pvalues <- function(superset){
  df <- summary(superset)$Table %>%
          as.data.frame() %>%
          dplyr::select(-Elements)
  return(df)
}




res_study35 <- supertest(study_36_contrasts, n = 20412)

res_study35_p <- extract_superset_pvalues(res_study35)

write_tsv(res_study35_p, "output/study36_chip_overlap_pvalues.txt")

pdf("output/figures/study36_chip-seq_overlap_test_circle_plot.pdf", width = 9)
plot(res_study35, sort.by="size", margin=c(2,2,2,8), color.scale.pos=c(0.85,1),
     legend.pos=c(0.9,0.15))
dev.off()


pdf("output/figures/study36_chip-seq_overlap_test_bar_plot.pdf", width = 9)
plot(res_study35, Layout="landscape", degree=2:4, sort.by="size", margin=c(0.5,8,1,2))
dev.off()




res_rv144 <- supertest(sign_gene_list, n = 20412)
res_rv144_p <- extract_superset_pvalues(res_rv144)

write_tsv(res_rv144_p, "output/rv144_chip_overlap_pvalues.txt")

pdf("output/figures/rv144_chip-seq_overlap_test_circle_plot.pdf", width = 9)
plot(res_rv144, sort.by="size", margin=c(2,8,2,8), color.scale.pos=c(0.85,1),
     legend.pos=c(0.9,0.15))
dev.off()


pdf("output/figures/rv144_chip-seq_overlap_test_bar_plot.pdf", width = 9)
plot(res_rv144, Layout="landscape", degree=2:4, sort.by="size", margin=c(0.5,15,1,6))
dev.off()











```


```{r study_rv_intersect}
combined_contrasts <- c(study_36_contrasts, sign_gene_list)
combined_contrasts[["CHIP-SeqCREB1 Peaks"]] <- NULL

res_combined <- supertest(combined_contrasts, n = 20412)

res_combined_p <- extract_superset_pvalues(res_combined)

write_tsv(res_combined_p, "output/study36_rv144_combined_chip_overlap_pvalues.txt")

pdf("output/figures/study36_rv144_combined_chip-seq_overlap_test_circle_plot.pdf", width = 14,
    height  = 14)
plot(res_combined, sort.by="size", margin=c(2,2,2,8), color.scale.pos=c(0.85,1),
     legend.pos=c(0.9,0.15))
dev.off()


combined_interactions <- summary(res_combined)$Table %>%
          rownames_to_column("overlap_code") %>%
          filter(grepl("CHIP", Intersections))  %>%
          .$overlap_code

pdf("output/figures/study36_rv144_combined_chip-seq_overlap_test_bar_plot.pdf", width = 7, height = 7)
plot(subset_mset(res_combined, Interactions = combined_interactions), Layout="landscape", degree=6:7, sort.by="size", margin=c(0.5,15,1,2))
dev.off()




b_interactions <- summary(res_nhp_chip_B)$Table %>%
          rownames_to_column("overlap_code") %>%
          filter(grepl("Study36", Intersections))  %>%
          filter(Degree >1) %>%
          .$overlap_code


pdf("output/figures/NHP_chip-seq_overlap_bar_plot_Bcells.pdf", width = 9)
plot(subset_mset(res_nhp_chip_B, Interactions = b_interactions), Layout="landscape", 
     degree=2:5, sort.by="size", margin=c(0.5,8,1,2))
dev.off()

combined_intersect <- Reduce(intersect, combined_contrasts)
writeLines(combined_intersect, con = "output/combined_chip_study36_rv144_intersect.txt")

```



```{r genemania_network_func}


import_genemania_network <- function(genemania_edgefile,genemania_nodefile, query_only = T ){
  net_in <- read_csv(genemania_edgefile) %>%
              dplyr::select(`data type`, `shared name`) %>%
              separate(., `shared name`, into = c("source_temp", "target_temp", "type"), 
                       sep = "[|]") 
  
  node_in <- read_csv(genemania_nodefile) %>%
              mutate(query_filter = ifelse(query_only == T & `node type` != "query", 0, 1)) %>%
              filter(query_filter == 1) %>%
              dplyr::select(`shared name`, `gene name`)
  net_out <- net_in %>%
              inner_join(., node_in, by = c("source_temp" = "shared name")) %>%
              dplyr::rename(source = `gene name`) %>%
              inner_join(., node_in, by = c("target_temp" = "shared name")) %>%
              dplyr::rename(target = `gene name`,
                            interaction = 'type') %>%
              dplyr::select(source, target, interaction) 
  return(net_out)

}

make_node_table <- function(network_table){
       out <- data.frame("node" = c(network_table$source, network_table$target)) %>%
              unique()
        return(out)                    
}


```



```{r rv144 genemania}

rv144_gm <- import_genemania_network(genemania_nodefile = 
                                       "output/rv144_overlap.genemania_node_raw.csv",
                         genemania_edgefile = "output/rv144_overlap.genemania_edge_raw.csv") %>%
              filter(!interaction %in% c("Predicted", 'Shared protein domains' ))
                    

rv144_node <- make_node_table(rv144_gm) %>%
                inner_join(., raw_contrasts, by = c("node" = "SYMBOL")) %>%
                dplyr::select(-immune_filter,-conditions) %>%
                inner_join(., filt_bed_full %>% dplyr::select(SYMBOL, V7), 
                           by = c("node" = "SYMBOL")) %>%
                group_by(node) %>%
                mutate(avg_fc = mean(V7)) %>%
                dplyr::select(-V7) %>%
                unique()

write_tsv(rv144_gm, path = "output/rv144_genemania.network_table.txt")
write_tsv(rv144_node, path = "output/rv144_genemania.node_table.txt")

```


```{r study36 genemania}


study36_gm <- import_genemania_network(genemania_nodefile = 
                                       "output/study36_overlap.genemania_node_raw.csv",
                         genemania_edgefile = "output/study36_overlap.genemania_edge_raw.csv") %>%
              filter(!interaction %in% c("Predicted", 'Shared protein domains' ))
                    

study36_node <- make_node_table(study36_gm) %>%
                inner_join(., study_36_genes, by = c("node" = "SYMBOL")) %>%
                dplyr::select(-immune_filter,-celltypes, -celltype) %>%
                unique() %>%
                inner_join(., filt_bed_full %>% dplyr::select(SYMBOL, V7), 
                           by = c("node" = "SYMBOL")) %>%
                group_by(node) %>%
                mutate(avg_fc = mean(V7)) %>%
                dplyr::select(-V7) %>%
                unique()

write_tsv(study36_gm, path = "output/study36_genemania.network_table.txt")
write_tsv(study36_node, path = "output/study36_genemania.node_table.txt")

```



```{r chip-seq_genesets}

make_intersect_geneset <- function(genelist, setname){
  
  out <- Reduce(intersect, genelist)
  
  
  return(GeneSet(out, setName = setname))
  
}


chip_intersect_coll <- list()
chip_intersect_coll <- list.append(chip_intersect_coll,
                                   make_intersect_geneset(sign_gene_list[c("RV144 HIV Acquisition",
                                                                   "CHIP-Seq CREB1 Peaks" )],
                                                  setname = "CHIP_RV144_acq"))

chip_intersect_coll <- list.append(chip_intersect_coll,
                                   make_intersect_geneset(sign_gene_list,
                                                  setname = "CHIP_RV144"))

chip_intersect_coll <- list.append(chip_intersect_coll,
                                   make_intersect_geneset(combined_contrasts,
                                                  setname = "CHIP_rv144_study36"))

toGmt(GeneSetCollection(chip_intersect_coll), con = "output/CHIP_transc_overlap.gmt")

```



```{r read_peaks}

peak_files_NHP <- list("NHP" = 
       list("Empty_24" = "NHP/NHP_run2/empty_vs_media_24hr/overlap.optimal_peak.narrowPeak.gz",
            "Full_24" = "NHP/NHP_run2/full_vs_media_24hr/overlap.optimal_peak.narrowPeak.gz",
            "Empty_48" = "NHP/NHP_run2/empty_vs_media_48hr/overlap.optimal_peak.narrowPeak.gz",
            "Full_48" = "NHP/NHP_run2/full_vs_media_48hr/overlap.optimal_peak.narrowPeak.gz")
       )


peak_calls_NHP <- sapply(names(peak_files_NHP), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_files_NHP[[x]]), simplify = F, USE.NAMES = T, function(y){
     out <- readPeakFile(peak_files_NHP[[x]][[y]])
   })
  
})


covplots_NHP <- sapply(names(peak_calls_NHP), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_calls_NHP[[x]]), simplify = F, USE.NAMES = T, function(y){
    out <- covplot(peak_calls_NHP[[x]][[y]], weightCol = "V7")
  })
})

jpeg("output/figures/NHP_empty_alvac_vs_media_24.jpg")
covplots_NHP$NHP$Empty_24
dev.off()
```


```{r promoters_NHP}



txdb_NHP <- TxDb.Mmulatta.UCSC.rheMac10.refGene
#promoter <- getPromoters(TxDb=txdb, upstream=1000, downstream=1000)

annot_matrix_NHP <- sapply(names(peak_calls_NHP), simplify = F, USE.NAMES = T, function(x){
  peak <- sapply(names(peak_calls_NHP[[x]]), simplify = F, USE.NAMES = T, function(y){
    out <- annotatePeak(peak_calls_NHP[[x]][[y]], TxDb =txdb_NHP, tssRegion = c(-3000,3000)) 
  })
})
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL")
mart_mmul = useMart("ensembl", dataset="mmulatta_gene_ensembl")


refseq_nhp <- unique(unlist(lapply(annot_matrix_NHP$NHP, function(x){
    out <- as.data.frame(x)$transcriptId })))



ensembl_id_NHP <-  getBM(attributes= c("refseq_mrna", "ensembl_gene_id"),
                  filters = 'refseq_mrna', values = refseq_nhp,
                  mart = mart_mmul)

hsapiens_orthologs <- getBM(attributes=
                              c("ensembl_gene_id","hsapiens_homolog_associated_gene_name"),
                  filters = 'ensembl_gene_id', values = ensembl_id_NHP$ensembl_gene_id,
                  mart = mart_mmul)


conversion_rate <- ensembl_id_NHP %>%
                    inner_join(., hsapiens_orthologs, by = "ensembl_gene_id") %>%
                    filter(ensembl_gene_id != "")


fraction_of_orthologs <- length(unique(conversion_rate$refseq_mrna))/ length(refseq_nhp)


annot_df_promoter_NHP <- sapply(names(annot_matrix_NHP), simplify = F, USE.NAMES = T, function(x){
    peak <- do.call("rbind", lapply(names(annot_matrix_NHP[[x]]), function(y){
      out <- as.data.frame(annot_matrix_NHP[[x]][[y]]) %>%
              mutate(species = x) %>%
              mutate(condition = y) %>%
              filter(grepl("Promoter", annotation)) #%>%
          #filter(grepl(paste(immune_filter, collapse = "|"),
                           #SYMBOL))
  })) 
}) 





prom_bed_full_NHP <- sapply(names(annot_matrix_NHP$NHP), simplify = F, USE.NAMES = T, function(x){
  df <- as.data.frame(annot_matrix_NHP$NHP[[x]]) %>%
              filter(grepl("Promoter", annotation))
  
  
})# %>%
              #filter(distanceToTSS <0) %>%

# 
# mart = useMart("ensembl", dataset="hsapiens_gene_ensembl")
# 
# 
# ensembl_id <-  getBM(attributes= c("hgnc_symbol","ensembl_gene_id", "strand"),
#                   filters = 'hgnc_symbol', values = prom_bed_full$SYMBOL, mart = mart) 


filt_bed_full_NHP <- sapply(names(prom_bed_full_NHP), simplify = F, USE.NAMES = T, function(x){
  
  df <- prom_bed_full_NHP[[x]] %>%
              dplyr::select(-strand) %>%
              #left_join(., ensembl_id, by = c("SYMBOL" = "hgnc_symbol")) %>%
              filter(distanceToTSS < 300) %>%
              filter(distanceToTSS > -3000) %>%
              inner_join(., ensembl_id_NHP, by = c("transcriptId" = "refseq_mrna")) %>%
              inner_join(., hsapiens_orthologs, by  = "ensembl_gene_id") %>%
              dplyr::rename(SYMBOL = "hsapiens_homolog_associated_gene_name") %>%
              filter(SYMBOL != "") %>%
              dplyr::select(seqnames, start, end, V7, SYMBOL, V4, V9) 
  
  
}) 


qc_NHP_peaks <- do.call("rbind",lapply(names(filt_bed_full_NHP), function(x){
  df <- filt_bed_full_NHP[[x]] %>%
          #filter(V7 >5 & V9 > -log2(0.05)) %>%
          filter( V9 > -log2(0.05)) %>%
          summarise(n = n()) %>%
          mutate(condition = x)
      
}))

filt_bed_full_NHP_qvals <- do.call("rbind",  sapply(names(prom_bed_full_NHP), simplify = F, USE.NAMES = T, function(x){
  
  df <- prom_bed_full_NHP[[x]] %>%
              dplyr::select(-strand) %>%
              #left_join(., ensembl_id, by = c("SYMBOL" = "hgnc_symbol")) %>%
              filter(distanceToTSS < 300) %>%
              filter(distanceToTSS >-3000) %>%
              inner_join(., ensembl_id_NHP, by = c("transcriptId" = "refseq_mrna")) %>%
              inner_join(., hsapiens_orthologs, by  = "ensembl_gene_id") %>%
              dplyr::rename(SYMBOL = "hsapiens_homolog_associated_gene_name") %>%
              filter(SYMBOL != "") %>%
              dplyr::select(seqnames, start, end, V9, SYMBOL, V4) %>%
              mutate(qval = 10^(-V9)) %>%
              filter(SYMBOL %in% c("CX3CL1", "HLA-E", "FLT3LG", "TBK1", "CCL2", "HLA-DMB")) %>%
              mutate(condition = x)
  
  
}) )
write_tsv(filt_bed_full_NHP_qvals, path = "output/suppl5C_peak_qvalues_NHP.txt")


lapply(names(filt_bed_full_NHP), function(x){
  write.table(filt_bed_full_NHP[[x]] %>% dplyr::select(-SYMBOL, -V4),
            file = file.path("NHP",paste(x, "filt_promoter_fc_pooled.bed", sep = "_")), 
            quote = F,
            col.names = F, row.names = F)

})
# 

venn.diagram(list("CHEA CREB1" = CREB_CHEA_LE, 
                  "NHP_Chip" = unique(unlist(NHP_creb_targets, use.names = F))),
             filename = "output/figures/venn_diagram_Full ALVAC_chip_vs_CHEA_NHP.tiff",
              fill = c(4,5),
             margin = 0.15, width = 3, height = 3, units = "in", resolution = 300,
             cex = 0.5, cat.cex = 0.5, cat.dist = 0.1)


NHP_creb_targets <- sapply(names(filt_bed_full_NHP), simplify = F, USE.NAMES = T, function(x){
  out <- filt_bed_full_NHP[[x]]$SYMBOL

})
# peak_calls_filt <- peak_calls$Human$Full[peak_calls$Human$Full$V4 %in% filt_bed_full$V4,]
# 
# covplot(peak_calls_filt, weightCol = "V7")
# 


CREB_CHEA <- getGmt("~/Documents/GMT_files/CHEA-tf.gmt")
CREB_CHEA_LE <- geneIds(CREB_CHEA[names(CREB_CHEA) == "CREB1"])[[1]]



putative_creb_raw <- read_excel("../transcriptomicss/input/01076Table3.xls", skip = 1) %>%
                      separate_rows(`Alias Symbols`, sep = ";")

alias <- putative_creb_raw$`Alias Symbols`[!is.na(putative_creb_raw$`Alias Symbols`)]
mart_m = useMart("ensembl", dataset="mmusculus_gene_ensembl")
# putat_orth_m <-  getBM(attributes= c("ensembl_gene_id", "external_gene_name"),
#                   filters = "external_gene_name", values = putative_creb_raw$Symbol, mart = mart_m)

putat_orth <-  getLDS(attributes= c("external_gene_name"),
                                   
                  filters = "external_gene_name", 
                  values = putative_creb_raw$Symbol, 
                             mart = mart_m,
                  attributesL= c("external_gene_name"),martL = mart) 

venn.diagram(NHP_creb_targets, fill = c(4,5,6,7),
             margin = 0.15, width = 3, height = 3, units = "in", resolution = 300,
             cex = 0.5, cat.cex = 0.5, cat.dist = 0.1, filename = "output/figures/NHP_overlap.tiff")




human_chip_overlap <- NHP_creb_targets
human_chip_overlap[["Human_chip_full_24"]] <- filt_bed_full$SYMBOL[!is.na(filt_bed_full$SYMBOL)]


res_nhp_chip <- supertest(human_chip_overlap, n = 21761) #ensembl
res_nhp_chip_p <- extract_superset_pvalues(res_nhp_chip)
pdf("output/figures/NHP_chip-seq_overlap_bar_plot.pdf", width = 9)
plot(res_nhp_chip, Layout="landscape", degree=2:4, sort.by="size", margin=c(0.5,8,1,2))
dev.off()


NHP_overlap <- c(NHP_creb_targets)

res_nhp_chip <- supertest(NHP_overlap, n = 21761) #ensembl
res_nhp_chip_p <- extract_superset_pvalues(res_nhp_chip)
pdf("output/figures/NHP_chip-seq_overlap_bar_plot.pdf", width = 9)
plot(res_nhp_chip, Layout="landscape", degree=2:5, sort.by="size", margin=c(0.5,8,1,2))
dev.off()


subset_mset <- function(mset, Interactions){
  mset_in <- mset
  
  mset_in[["overlap.sizes"]] <- mset$overlap.sizes[names(mset$overlap.sizes) %in% Interactions]
  mset_in[["overlap.expected"]] <- mset$overlap.expected[names(mset$overlap.expected) %in%
                                                           Interactions]
  
  mset_in[["P.value"]] <- mset$P.value[names(mset$P.value) %in%
                                                           Interactions]
  return(mset_in)
}



study36_b <- c(NHP_creb_targets, list("Study36 B" = unique(study_36_genes[study_36_genes$celltype ==
                                                           "Bcells",]$SYMBOL)))

res_nhp_chip_B <- supertest(study36_b, n = 21761) #ensembl
res_nhp_chip_B_p <- extract_superset_pvalues(res_nhp_chip_B)


b_interactions <- summary(res_nhp_chip_B)$Table %>%
          rownames_to_column("overlap_code") %>%
          filter(grepl("Study36", Intersections))  %>%
          filter(Degree >1) %>%
          .$overlap_code


pdf("output/figures/NHP_chip-seq_overlap_bar_plot_Bcells.pdf", width = 9)
plot(subset_mset(res_nhp_chip_B, Interactions = b_interactions), Layout="landscape", 
     degree=2:5, sort.by="size", margin=c(0.5,8,1,2))
dev.off()



study36_T <- c(NHP_creb_targets, list("Study36 T" = unique(study_36_genes[study_36_genes$celltype ==
                                                           "Tcells",]$SYMBOL)))
res_nhp_chip_T <- supertest(study36_T, n = 21761) #ensembl
res_nhp_chip_T_p <- extract_superset_pvalues(res_nhp_chip_T)
t_interactions <- summary(res_nhp_chip_T)$Table %>%
          rownames_to_column("overlap_code") %>%
          filter(grepl("Study36", Intersections))  %>%
          filter(Degree >1) %>%
          .$overlap_code
pdf("output/figures/NHP_chip-seq_overlap_bar_plot_Tcells.pdf", width = 9)
plot(subset_mset(res_nhp_chip_T, t_interactions), Layout="landscape", degree=2:5, 
     sort.by="size", margin=c(0.5,8,1,2))
dev.off()


study36_DC <- c(NHP_creb_targets, list("Study36 DC" = unique(study_36_genes[study_36_genes$celltype == "DCcells",]$SYMBOL)))

res_nhp_chip_DC <- supertest(study36_DC, n = 21761) #ensembl
res_nhp_chip_DC_p <- extract_superset_pvalues(res_nhp_chip_DC)
dc_interactions <- summary(res_nhp_chip_DC)$Table %>%
          rownames_to_column("overlap_code") %>%
          filter(grepl("Study36", Intersections))  %>%
          filter(Degree >1) %>%
          .$overlap_code
pdf("output/figures/NHP_chip-seq_overlap_bar_plot_DCcells.pdf", width = 9)
plot(subset_mset(res_nhp_chip_DC, dc_interactions), Layout="landscape", degree=2:5,
     sort.by="size", margin=c(0.5,8,1,2))
dev.off()






NHP_gene_summary_temp <- do.call("rbind", lapply(names(study_36_contrasts)[c(1:3)], function(x){
    out <- data.frame("SYMBOL" = study_36_contrasts[[x]]) %>%
              mutate(study36_cellPop_solo = x) %>%
              unique()
})) %>%
  group_by(SYMBOL) %>%
  mutate(study36_cellPop = paste(study36_cellPop_solo, collapse = "; ")) %>%
  dplyr::select(-study36_cellPop_solo) %>%
  unique()


NHP_gene_summary <- do.call("rbind", lapply(names(NHP_creb_targets), function(x){
    out <- data.frame("SYMBOL" = NHP_creb_targets[[x]]) %>%
              mutate(condition_solo = x) %>%
              unique()
})) %>%
  group_by(SYMBOL) %>%
  mutate(conditions = paste(condition_solo, collapse = "; ")) %>%
  dplyr::select(-condition_solo) %>%
  unique() %>%
  left_join(., NHP_gene_summary_temp, by = "SYMBOL") %>%
  mutate(immune_filter = ifelse(grepl(paste(immune_filter, collapse = "|"),SYMBOL),
                                "YES", "NO")) %>%
  separate_rows(conditions, sep = "; ") %>%
  mutate(condition_value = "X") %>%
  spread(conditions,  condition_value) %>%
  gather(conditions, condition_value, -SYMBOL, -study36_cellPop, -immune_filter) %>%
  mutate(condition_value = ifelse(is.na(condition_value), "", condition_value)) %>%
  spread(conditions,  condition_value) %>%
  separate_rows(study36_cellPop, sep = "; ") %>%
  mutate(cell_pop_value = ifelse(!is.na(study36_cellPop), "X", NA)) %>%
  spread(study36_cellPop, cell_pop_value) %>%
  gather(study36_cellPop, cell_pop_value, names(study_36_contrasts)[c(1:3)] ) %>%
  
  mutate(cell_pop_value = ifelse(is.na(cell_pop_value ), "",cell_pop_value )) %>%
  spread(study36_cellPop, cell_pop_value) %>%
  dplyr::select(-7)
 

immune_filter <- c('CCL', 'CXCL', 'CX3CL', "CCR", 'CXCR', "CX3CR", "MIP",
                   'DDX', 'TMEM', 'IL[0-9]+', 'IFN', 'TGF' , 'FLT3L', 'TBK', 'TNF',
                   '^CD[0-9]+', 'IRF', 'CDKN', 'HLA', 'FOS', 'JUN', 'TLR', 'MAPK', 'NOD', 'NLR',
                   'NFKB', 'REL[ABC]', "STAT")



```



```{r make Mmul_txDb}
library(GenomicFeatures)
library(AnnotationDbi)


metadata_Mmul_10 <- data.frame("name" = c("source", "build"),
                            "value" =
                          c("https://hgdownload.soe.ucsc.edu/hubs/GCF/003/339/765/GCF_003339765.1/",
                            "Mmul_10"))
Mmul_10_txDB <- makeTxDbFromGFF("NHP/NHP_run2/genes.gtf",
                                organism = "Macaca mulatta",
                                metadata = metadata_Mmul_10)

#saveDb(Mmul_10_txDB, file = "output/Mmul_10_TxDb.RDS")
```

```{r overlap_qc}

qcLS <- list.files("NHP/NHP_run2/", pattern = ".*overlap.*qc", 
                   full.names = T, recursive = T)

qcLS <- qcLS[!grepl("pseudo", qcLS)]
qcLS <- qcLS[grepl("bfilt", qcLS)]

qcDF <- sapply(qcLS, simplify = F, USE.NAMES = T,function(x){
    df <- read.table(x)
  
  
})

```
